### PR TITLE
[WIP] Web API + web API test

### PR DESCRIPTION
**Pull request by Heimdell**
## Description

Web API for Coffer + js mocha tests.

## Related issue(s)

None

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

- Public contracts
  - [ ] Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - [ ] I added an entry to the [changelog](../tree/master/CHANGES.md) if my changes are visible to the users
        and
  - [ ] provided a migration guide for breaking changes if possible

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: tests/api/test.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
+let {GET, POST, DELETE, all, contains, json_equal, rand, noError, explode } = require("./utils")
```
Comment (DK318) [inline tests/api/test.js:1]: I don't like that these tests are depending on previous ones. I think we should do smth like `coffer delete / -r` before running each test (like in `bats` tests). + you can look [here](https://github.com/serokell/coffer/pull/32#discussion_r820250737) and write missing tests.

**Review by DK318 — CHANGES_REQUESTED** [review]

File: lib/Web/API.hs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+          void "date"
+          op <- choice
+            [ ">=" $> OpGTE
+            , "<=" $> OpLTE
+            , ">"  $> OpGT
+            , "<"  $> OpLT
+            , "="  $> OpEQ
+            ]
+          FilterByDate op <$> parseDate
+      ]
```
[the remark this answers is not in the document]
Comment (Heimdell) [reply lib/Web/API.hs:112]: Guess I mislooked that concrete one, will reuse if possible. Internal `MParser` of `optparse-applicative` is not runnable outside.

File: lib/Web/API.hs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+          ]
+
+parseDate :: Parsec Void Text FilterDate
+parseDate = parseFilterDate
+
+digit :: Parsec Void Text Char
+digit = oneOf ['0'..'9'] <?> "digit"
+
+times :: Applicative m => Int -> m a -> m [a]
+times = (sequenceA .) . replicate
```
[the remark this answers is not in the document]
Comment (Heimdell) [reply lib/Web/API.hs:173]: Yes, will replace

File: tests/api/test.js (hunk 1 above)
Comment (Heimdell) [reply tests/api/test.js:1]: Will do. I wondered how can I clear the state entirely, thanks!

Comment (dcastro) [reply tests/api/test.js:1]: > + you can look here and write missing tests

I don't think we need to test every edge case for the Web API as we did for the CLI. Since they use the same core modules, that seems like duplicated effort.

We should treat it as "just another interface" for coffer's core, and test it as such: 
* make sure all inputs (URI path, query params, headers, etc) are correctly parsed and processed
* make sure all outputs (including error scenarios) are correctly formatted

Testing the output of error scenarios does involve testing a lot of edge cases, but perhaps not as many as we did in the CLI, I hope?

**Review by dcastro — CHANGES_REQUESTED** [review]

File: lib/Web/API.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dcastro) [reply lib/Web/API.hs:112]: Created an issue to improve this after #37 is merged: #89

**Comment by dcastro** [comment]
@DK318: Can you please take over this PR?

Here's the plan:
1. Rebase it on main and fix any conflicts
1. Reorganize commits into atomic changes. Right now there's a lot of commits that merely fix or amend previous commits (e.g. "Fixes after rebase", "Add /find test"). I think it would make sense to have these commits (just a suggestion):
    * Implement Web API (`Server.hs` and `API.hs`)
    * Add an executable component for the Web API (edits `package.yaml`, moves the CLI's `cli/Main.hs`, adds a new `server/Main.hs`).
    * Add mocha tests for the Web API
1. Add Problem/Solution to each commit, tag each commit with the issue number (I created an issue here: #93)
1. Get the pipeline green
1. Address the comments you and I left on this PR

If any concerns arise, please open a new issue and we'll fix it after #37 is merged. I already opened a few and tagged them with the `web-api` label, see: https://github.com/serokell/coffer/labels/web-api . This will allow the team to work on them in parallel.

Unless it's a critical issue, in which case we should talk about it and possibly solve it here in #37.

File: tests/api/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+node_modules
```
[the remark this answers is not in the document]
Comment (DK318) [reply tests/api/.gitignore]: Done

File: lib/Web/API.hs [hunk 5]
```diff
@@ -0,0 +1,182 @@
+{-# options_ghc -Wno-orphans #-}
```
[the remark this answers is not in the document]
Comment (DK318) [reply lib/Web/API.hs]: Done

File: lib/Web/API.hs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+{-# OPTIONS_GHC -Wno-deferred-out-of-scope-variables #-}
```
[the remark this answers is not in the document]
Comment (DK318) [reply lib/Web/API.hs]: Done

**Comment by sancho20021** [comment]
Some new conflicts appeared, as there were recent pushes in main branch :( Would you please resolve them?

**Comment by DK318** [comment]
@sancho20021 I'll resolve them when I start merging

File: lib/CLI/Types.hs [hunk 7]
```diff
@@ -168,34 +210,170 @@ data FieldInfo = FieldInfo
...
+  where
+    -- | Parse a two-digit integer (e.g. day of month, hour).
+    twoDigits :: MParser Int
+    twoDigits = do
+      a <- P.digitChar
+      b <- P.digitChar
+      pure $ Char.digitToInt a * 10 + Char.digitToInt b
+
+digit :: MParser Char
+digit = oneOf ['0'..'9'] <?> "digit"
```
Comment (dcastro) [inline lib/CLI/Types.hs:376]: This doesn't seem to be needed.

File: .reuse/dep5 [hunk 8]
```diff
@@ -1,5 +1,5 @@
 Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
 
-Files: .github/pull_request_template.md .github/ISSUE_TEMPLATE/*.md
+Files: .github/pull_request_template.md .github/ISSUE_TEMPLATE/*.md *.json
```
Comment (dcastro) [inline .reuse/dep5:3]: Excluding all json files may be too lax, can we exclude just specific files?

File: lib/Coffer/Path.hs [hunk 9]
```diff
@@ -177,14 +196,30 @@ data QualifiedPath path = QualifiedPath
...
+  parseUrlPiece text
+    | [pathPiece] <- T.splitOn "#" text = do
+        path <- parseUrlPiece @path pathPiece
+        pure $ QualifiedPath Nothing path
+    | [backendPiece, pathPiece] <- T.splitOn "#" text = do
+        backendName <- newBackendName backendPiece
+        path <- parseUrlPiece @path pathPiece
+        pure $ QualifiedPath (Just backendName) path
+    | otherwise =
+        Left "Invalid qualified path format. Expected [BACKENDNAME#]PATH"
```
Comment (dcastro) [inline lib/Coffer/Path.hs:218]: This could be refactored to avoid using `T.splitOn "#"` twice. But don't worry about it, I've added a note in #89 to refactor it later.

File: .reuse/dep5 (hunk 8 above)
Comment (DK318) [reply .reuse/dep5:3]: We can't use comments in `json` AFAIK

Comment (dcastro) [reply .reuse/dep5:3]: You can have a separate file named `filename.json.license` which contains the license info for `filename.json`.
See `nix/sources.json.license` as an example.

File: lib/CLI/Types.hs (hunk 7 above)
Comment (DK318) [reply lib/CLI/Types.hs]: Removed

File: .reuse/dep5 (hunk 8 above)
Comment (DK318) [reply .reuse/dep5]: Done

**Review by dcastro — APPROVED** [review]

File: lib/CLI/Types.hs [hunk 10]
```diff
@@ -167,35 +209,168 @@ data FieldInfo = FieldInfo
...
+          void "date"
+          op <- choice
+            [ ">=" $> OpGTE
+            , "<=" $> OpLTE
+            , ">"  $> OpGT
+            , "<"  $> OpLT
+            , "="  $> OpEQ
+            ]
+          FilterByDate op <$> parseFilterDate
+      ]
```
Comment (sancho20021) [inline lib/CLI/Types.hs:274]: Is the handling of the third constructor, `FilterByField`, is omitted here on purpose?

File: lib/Web/Server.hs [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    :<|> set     run token txt fkey
+    )
+  :<|> deleteField run token
+  :<|> find'       run token
+  :<|> rename      run token
+  :<|> copy'       run token
+  :<|> delete'     run token
+  :<|> tag         run token
+
+view
```
Comment (sancho20021) [inline lib/Web/Server.hs:93]: I'm not sure, but could the common repeating code
```
_ run token _ = run token _
``` 
be extracted and written once? (In the next group of functions)

File: lib/CLI/Types.hs (hunk 10 above)
Comment (DK318) [reply lib/CLI/Types.hs:274]: Hmm. I think I forgot to handle it after rebasing. But I think it could be solved when we implement #89

File: lib/Web/Server.hs (hunk 11 above)
Comment (DK318) [reply lib/Web/Server.hs:93]: I think it wouldn't type check

File: lib/CLI/Types.hs (hunk 10 above)
Comment (sancho20021) [reply lib/CLI/Types.hs:274]: Agree, we just have not to forget about this. Maybe put some todo comment?

File: lib/Web/Server.hs (hunk 11 above)
Comment (sancho20021) [reply lib/Web/Server.hs:93]: Ah, I see
